atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/expectations/TrackingKeyValueServiceReadInfoTest.java

```diff
@@ -0,0 +1,321 @@
+        Map<RangeRequest, TokenBackedBasicResultsPage<RowResult<Value>, byte[]>> pageByRangeRequestMapOfSize =
+                TrackingKeyValueServiceTestUtils.createPageByRangeRequestMapWithSize(size);
+
+        when(kvs.getFirstBatchForRanges(tableReference, rangeRequests, TIMESTAMP))
+                .thenReturn(pageByRangeRequestMapOfSize);
+
+        trackingKvs.getFirstBatchForRanges(tableReference, rangeRequests, TIMESTAMP);
+
+        validateReadInfoForReadForTable("getFirstBatchForRanges");
+    }
+
+    @Test
+    public void readInfoIsCorrectAfterGetAllTableNamesCall() {
+        when(kvs.getAllTableNames()).thenReturn(TrackingKeyValueServiceTestUtils.createTableReferenceSetWithSize(size));
+
+        trackingKvs.getAllTableNames();
+
+        validateReadInfoForTableAgnosticRead("getAllTableNames");
+    }
+
+    @Test
+    public void readInfoIsCorrectAfterGetMetadataForTableCall() {
+        when(kvs.getMetadataForTable(tableReference)).thenReturn(new byte[size]);
+
+        trackingKvs.getMetadataForTable(tableReference);
+
+        validateReadInfoForTableAgnosticRead("getMetadataForTable");
+    }
+
+    @Test
+    public void readInfoIsCorrectAfterGetMetadataForTablesCall() {
+        Map<TableReference, byte[]> metadataForTablesOfSize =
+                TrackingKeyValueServiceTestUtils.createByteArrayByTableReferenceMapWithSize(size);
+
+        when(kvs.getMetadataForTables()).thenReturn(metadataForTablesOfSize);
+
+        trackingKvs.getMetadataForTables();
+
+        validateReadInfoForTableAgnosticRead("getMetadataForTables");
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void readInfoIsCorrectAfterGetAllTimestampsCall() {
+        Set<Cell> cells = mock(Set.class);
+        Multimap<Cell, Long> longByCellMultimapOfSize =
+                TrackingKeyValueServiceTestUtils.createLongByCellMultimapWithSize(size);
+
+        when(kvs.getAllTimestamps(tableReference, cells, TIMESTAMP)).thenReturn(longByCellMultimapOfSize);
+
+        trackingKvs.getAllTimestamps(tableReference, cells, TIMESTAMP);
+
+        validateReadInfoForReadForTable("getAllTimestamps");
+    }
+
+    @Test
+    public void readInfoIsCorrectAfterGetRowsKeysInRangeCall() {
+        int maxResults = 13;
+        byte[] startRow = new byte[1];
+        byte[] endRow = new byte[3];
+        ImmutableList<byte[]> byteArrayListOfSize =
+                TrackingKeyValueServiceTestUtils.createListOfByteArraysWithSize(size);
+
+        when(kvs.getRowKeysInRange(tableReference, startRow, endRow, maxResults))
+                .thenReturn(byteArrayListOfSize);
+
+        trackingKvs.getRowKeysInRange(tableReference, startRow, endRow, maxResults);
+
+        validateReadInfoForReadForTable("getRowKeysInRange");
+    }
+
+    private void validateReadInfoForReadForTable(String methodName) {
+        TransactionReadInfo readInfo = ImmutableTransactionReadInfo.builder()
+                .bytesRead(size)
+                .kvsCalls(1)
```

Comment:
do any tests go beyond a single kvs call?

Reply:
No, I could add more tests for that. However, this is already kind of tested in `KeyValueServiceDataTracker` which is what the `TrackingKeyValueServiceImpl` uses to track.

These tests are meant to check that each method does somehow track (i.e. correctly delegate to the data tracker). Admittedly, there is some coupling here (and the testing strategy is implementation dependent); so I am open to different ways of doing this!